framework/src/bcs/LowerDIntegratedBC.C

```diff
@@ -34,15 +34,15 @@ LowerDIntegratedBC::LowerDIntegratedBC(const InputParameters & parameters)
     _test_lambda(_lowerd_var.phiLower())
 {
   const auto & lower_domains = _lowerd_var.activeSubdomains();
-  if (!lower_domains.count(Moose::BOUNDARY_SIDE_LOWERD_ID) && lower_domains.size() != 1)
-    paramError(
-        "lowerd_variable",
-        "Must be only defined on the subdomain BOUNDARY_SIDE_LOWERD_SUBDOMAIN subdomain that is "
-        "added by Mesh/build_all_side_lowerd_mesh=true");
-
-  if (_var.activeSubdomains().count(Moose::BOUNDARY_SIDE_LOWERD_ID))
-    paramError("variable",
-               "Must not be defined on the subdomain INTERNAL_SIDE_LOWERD_SUBDOMAIN subdomain");
+  for (const auto & id : _mesh.boundaryLowerDBlocks())
```

Comment:
The logic has to be this way. The variable on lower-d mesh actually can be defined both on the boundary and interior. In such a case, `lower_domains` will contain a block id that not in `_mesh.boundaryLowerDBlocks()`. The check here is to make sure the lower-d variable is valid on all boundary lower-d mesh blocks.

Reply:
Ok let’s check with the two sets then

there is no reason to force the variable to be defined on all lowerD blocks

the intent of this check is to prevent the variable from being defined on a higher D block. No need to change that.

Reply:
>let’s check with the two sets then

What do you mean? Checking this lower-d variable is defined on all boundary lower-d blocks is too much but that is consistent with the current code, because the lower-d block is covering the entire boundary sides.

More rigorous check should be making sure the boundary sides in `boundary` parameter of this BC are covered by the variable on the corresponding lower-d block, but we currently do not have a clean way to do that check.

Theoretically speaking, this variable can be defined on a higher-d block at the same time.

Reply:
This is pretty mind boggling...

The old logic is
```
  if (!lower_domains.count(Moose::BOUNDARY_SIDE_LOWERD_ID) && lower_domains.size() != 1)
    paramError(
        "lowerd_variable",
        "Must be only defined on the subdomain BOUNDARY_SIDE_LOWERD_SUBDOMAIN subdomain that is "
        "added by Mesh/build_all_side_lowerd_mesh=true");
```
which means when the lower-d variable only has ONE block, it will never hit this error, no matter what that block is. So this check is actually not right. What we want is to make sure this lower-d variable is defined on ALL boundary lower-d blocks. And the variable is not necessarily only on the boundary lower-d blocks. Let us say the variable is indeed on multiple blocks that makes this check meaningful, this logic is indeed to make sure the variable is defined on all boundary lower-d blocks.

The current logic
```
  const auto & lower_domains = _lowerd_var.activeSubdomains();
  for (const auto & id : _mesh.boundaryLowerDBlocks())
    if (lower_domains.count(id) == 0)
      paramError("lowerd_variable",
                 "Must be defined on the boundary lower-dimensional subdomain '" +
                     _mesh.getSubdomainName(id) +
                     "' that is added by Mesh/build_all_side_lowerd_mesh=true");
```
makes sure the variable is defined on the boundary lower-d block, and the code is consistent with the error message.

Reply:
Not need all, but for now, this is required. When the variable is on multiple blocks, this check is consistent with the old check. This is a BC, not on all blocks. But currently we require the lower-d variable is defined on all boundary blocks.

Reply:
The current check does not impose for the variable to be defined on all lower D blocks. It imposes that the variable is at least defined on `Moose::BOUNDARY_SIDE_LOWERD_ID` though in a weird way

> This is a BC, not on all blocks.

you made the same mistake in the kernels
```
  for (const auto & id : _mesh.interiorLowerDBlocks())
    if (lower_domains.count(id) == 0)
      paramError("lowerd_variable",
                 "Must be defined on the interior lower-dimensional subdomain '" +
                     _mesh.getSubdomainName(id) +
                     "'that is added by Mesh/build_all_side_lowerd_mesh=true");
```

Reply:
>it imposes that the variable is at least defined on Moose::BOUNDARY_SIDE_LOWERD_ID

Yes, that will guarantee this BC will work properly.

Reply:
The same for the kernels, if the lower-d variable is defined on all lower-d interior blocks, the kernel will work properly, otherwise, it may not because there could be an interior side where the variable does not have definition.

Reply:
Added an issue for overly restrict check here: https://github.com/idaholab/moose/issues/29151